### PR TITLE
EfficientUNet++, ResUnet and ResUnet++

### PR DESCRIPTION
**Pull request by jlcsilva**
Implemented the EfficientUNet++ (authored by me - https://arxiv.org/abs/2106.11447), the ResUnet and ResUnet++

### CODE REVIEW

**Comment by PhilippMarquardt** [comment]
Hi, EfficientUnet++ and ResUnet seem to be working fine but ResUnet++ gives an error: `DecoderBlock' object has no attribute 'attention2'`. I've installed your fork via pip cause I'm inetereset in testing the new decoders out.

**Comment by PhilippMarquardt** [comment]
Hi, I think you've missed adding `self.attention2 = md.Attention(attention_type, in_channels=out_channels)` [here](https://github.com/jlcsilva/segmentation_models.pytorch/blob/53c7f956ca557eb2cf386d28faacadf30ce4d0e2/segmentation_models_pytorch/resunetplusplus/decoder.py#L114). Currently it fails

**Review by jlcsilva — COMMENTED** [review]
Mean ResUNet++, not ResUNet

**Comment by jlcsilva** [comment]
> Hi, I think you've missed adding `self.attention2 = md.Attention(attention_type, in_channels=out_channels)` [here](https://github.com/jlcsilva/segmentation_models.pytorch/blob/53c7f956ca557eb2cf386d28faacadf30ce4d0e2/segmentation_models_pytorch/resunetplusplus/decoder.py#L114). Currently it fails to run ResUnet++ as decoder

You're absolutely right. It should be working now. 

If you're interested in the EfficientUNet++, check out this repo [repo](https://github.com/jlcsilva/EfficientUNetPlusPlus), where I briefly explain the architecture and compare its efficiency with other decoders'. For full details you can check the [paper](https://arxiv.org/abs/2106.11447)

So far, I've only tested it on a coronary artery segmentation task, but I think performance should hold for other tasks. Let me know if it works well for you

**Comment by PhilippMarquardt** [comment]
I've also noticed that the loss for my dataset gets oftentimes nan for epochs when using the ResUnet++ with this version. Using the ResUnet++ decoder from your `new_decoders` branch works fine on the other hand. All other decoders work fine as well without any nan's occuring. Have you had this problem as well?

**Comment by jlcsilva** [comment]
I didn't remember, but they're actually different versions. In the new_decoders branch, [this line](https://github.com/qubvel/segmentation_models.pytorch/blob/921c8d63ac0dcdfbeccedef0ce7afccc0268ff81/segmentation_models_pytorch/resunetplusplus/decoder.py#L95) is commented out. 

At that time, I was using a very small dataset (30 images), and attention was causing it to underperform (don't remember if it was overfitting or not even converging). Later, I got access to a larger dataset (~230 images) and decided to test attention once again. It turns out that with more data it performed very well (although very slowly compared to the EfficientUNet++ and UNet++).

If you comment out the attention, it will probably work fine. Whether to use attention or not could possibly be an argument to the class constructor, but I guess that in that case you could probably use the ResUNet instead.

However, the NaNs indicate you probably have a problem with your loss function. The output of the ResUNet++ should be causing the loss function to make a division by 0. You can solve that with smoothing. E.g., if your loss function is something like loss = x / y, you should use loss = x / (y + eps) instead, with eps=0.0001 or any other small number that doesn't affect the value of the loss but prevents division by 0

**Comment by cwerner** [comment]
Hi @jlcsilva and all! :wave:

I'm interested in trying EfficientUNet++ for a remote sensing application and was wondering if anything major is blocking this PR or if I should rather go use the fork at https://github.com/jlcsilva/EfficientUNetPlusPlus for the moment? Anyone applied these models outside of medical image segmentations?

Thanks!

**Comment by jlcsilva** [comment]
Hey @cwerner!

To the best of my knowledge, there are no issues. I think the maintainers simply haven't had time to review it yet. 

The fork at [https://github.com/jlcsilva/EfficientUNetPlusPlus](https://github.com/jlcsilva/EfficientUNetPlusPlus) actually uses this one as a submodule, so the model is the same. Using one or the other depends on your preference. If you already have a training pipeline for segmentation, it is probably easier to use this fork. Otherwise, you can use the training pipeline I made available in the other repo. It is very simple and can certainly be improved, but I think it will allow you to get the job done 😃 

A friend of mine is using the model in a drone vision system to detect zones of higher salt concentration in the ocean (which usually have a slightly different color) and he is getting good results. Apart from that, I don't know of any other applications outside the scope of medical image segmentation. Nevertheless, I'm almost sure it will work well in other domains. The ResUNet, for example, was originally proposed for road extraction, and had very good results, proving the versatility of UNet-based models.  The only thing I'm not sure about is how it will behave in the presence of a high number of classes, but I think it should be fine as well. 

Feel free to contact me if you have any issues with it or need any help. And please let me know how it performs in your task!

**Comment by JulienMaille** [comment]
@qubvel would you consider reopening that PR?
